python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -0,0 +1,152 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Partitioned LogicalPlan nodes."""
+
+from __future__ import annotations
+
+from functools import singledispatch
+from typing import TYPE_CHECKING, Any
+
+from cudf_polars.dsl.expr import NamedExpr
+from cudf_polars.dsl.traversal import reuse_if_unchanged, traversal
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.containers import DataFrame
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.dsl.nodebase import Node
+
+
+class PartitionInfo:
+    """
+    Partitioning information.
+
+    This class only tracks the partition count (for now).
+    """
+
+    __slots__ = ("count",)
+
+    def __init__(self, count: int):
+        self.count = count
+
+
+# The hash of an IR object must always map to a
+# unique PartitionInfo object, and we can cache
+# this mapping until evaluation is complete.
+_IR_PARTS_CACHE: MutableMapping[int, PartitionInfo] = {}
+
+
+def _clear_parts_info_cache() -> None:
+    """Clear cached partitioning information."""
+    _IR_PARTS_CACHE.clear()
+
+
+def get_key_name(node: Node | NamedExpr) -> str:
+    """Generate the key name for a Node."""
+    if isinstance(node, NamedExpr):
+        return f"named-{get_key_name(node.value)}"  # pragma: no cover
+    return f"{type(node).__name__.lower()}-{hash(node)}"
+
+
+@singledispatch
+def lower_ir_node(ir: IR, rec) -> IR:
+    """Rewrite an IR node with proper partitioning."""
+    # Return same node by default
+    return reuse_if_unchanged(ir, rec)
+
+
+def lower_ir_graph(ir: IR) -> IR:
+    """Rewrite an IR graph with proper partitioning."""
+    from cudf_polars.dsl.traversal import CachingVisitor
+
+    mapper = CachingVisitor(lower_ir_node)
+    return mapper(ir)
+
+
+def _default_ir_parts_info(ir: IR) -> PartitionInfo:
+    # Single-partition default behavior.
+    # This is used by `_ir_parts_info` for all unregistered IR sub-types.
+    count = max((ir_parts_info(child).count for child in ir.children), default=1)
+    if count > 1:
+        raise NotImplementedError(
+            f"Class {type(ir)} does not support multiple partitions."
+        )  # pragma: no cover
+    return PartitionInfo(count=count)
+
+
+@singledispatch
+def _ir_parts_info(ir: IR) -> PartitionInfo:
+    """IR partitioning-info dispatch."""
+    return _default_ir_parts_info(ir)
+
+
+def ir_parts_info(ir: IR) -> PartitionInfo:
```

Comment:
@wence- - I moved all partitioning logic out of `ir.py`. So, the partitioning information is now dispatched (and cached) when we call `ir_parts_info(ir)`. This dispatch function will return `PartitionInfo(count=1)` for all vanilla IR nodes. When we introduce new "multi-partition" nodes, the dispatched  `_ir_parts_info` logic will need to follow deterministic rules to calculate the partitioning (depending **only** on IR constructor arguments).

Reply:
**Note**: I'm calling this `ir_parts_info`, because we will also need something like `expr_parts_info(expr: Expr, child_ir: IR) -> PartitionInfo` to handle the full spectrum of `Select` logic.

Reply:
:thu